Mathlib/CategoryTheory/ChosenFiniteProducts.lean

```diff
@@ -195,6 +195,247 @@ instance (priority := 100) : Limits.HasFiniteProducts C :=
+@[reassoc]
+theorem prodComparison_inv_natural (f : A ⟶ A') (g : B ⟶ B') [IsIso (prodComparison F A' B')] :
+    inv (prodComparison F A B) ≫ F.map (f ⊗ g) =
+      (F.map f ⊗ F.map g) ≫ inv (prodComparison F A' B') := by
+  rw [IsIso.eq_comp_inv, Category.assoc, IsIso.inv_comp_eq, prodComparison_natural]
+
+/-- If the product comparison morphism is an iso, its inverse is natural in the right argument. -/
+@[reassoc]
+theorem prodComparison_inv_natural_whiskerLeft (g : B ⟶ B') [IsIso (prodComparison F A B')] :
+    inv (prodComparison F A B) ≫ F.map (A ◁ g) =
+      (F.obj A ◁ F.map g) ≫ inv (prodComparison F A B') := by
+  rw [IsIso.eq_comp_inv, Category.assoc, IsIso.inv_comp_eq, prodComparison_natural_whiskerLeft]
+
+/-- If the product comparison morphism is an iso, its inverse is natural in the left argument. -/
+@[reassoc]
+theorem prodComparison_inv_natural_whiskerRight (f : A ⟶ A') [IsIso (prodComparison F A' B)] :
+    inv (prodComparison F A B) ≫ F.map (f ▷ B) =
+      (F.map f ▷ F.obj B) ≫ inv (prodComparison F A' B) := by
+  rw [IsIso.eq_comp_inv, Category.assoc, IsIso.inv_comp_eq, prodComparison_natural_whiskerRight]
+
+end
+
+/-- The product comparison morphism from `F(A ⊗ -)` to `FA ⊗ F-`, whose components are given by
+`prodComparison`. -/
+@[simps]
+def prodComparisonNatTrans (A : C) :
+    (curriedTensor C).obj A ⋙ F ⟶ F ⋙ (curriedTensor D).obj (F.obj A) where
+  app B := prodComparison F A B
+  naturality x y f := by
+    apply hom_ext <;>
+    simp only [Functor.comp_obj, curriedTensor_obj_obj,
+      Functor.comp_map, curriedTensor_obj_map, Category.assoc, prodComparison_fst, whiskerLeft_fst,
+      prodComparison_snd, prodComparison_snd_assoc, whiskerLeft_snd, ← F.map_comp]
+
+/-- The product comparison morphism from `F(- ⊗ -)` to `F- ⊗ F-`, whose components are given by
+`prodComparison`. -/
+@[simps]
+def prodComparisonBifunctorNatTrans :
+    curriedTensor C ⋙ (whiskeringRight _ _ _).obj F ⟶
+      F ⋙ curriedTensor D ⋙ (whiskeringLeft _ _ _).obj F where
+  app A := prodComparisonNatTrans F A
+  naturality x y f := by
+    ext z
+    apply hom_ext <;>
+    simp only [Functor.comp_obj, whiskeringRight_obj_obj, curriedTensor_obj_obj,
+      whiskeringLeft_obj_obj, Functor.comp_map, whiskeringRight_obj_map, NatTrans.comp_app,
+      whiskerRight_app, curriedTensor_map_app, prodComparisonNatTrans_app, Category.assoc,
+      prodComparison_fst, whiskeringLeft_obj_map, whiskerLeft_app, whiskerRight_fst,
+      prodComparison_fst_assoc, prodComparison_snd, whiskerRight_snd, ← F.map_comp]
+
+/-- The natural isomorphism `F(A ⊗ -) ≅ FA ⊗ F-`, provided each `prodComparison F A B` is an
+isomorphism (as `B` changes). -/
+@[simps]
+noncomputable def prodComparisonNatIso (A : C)
+    [∀ B, IsIso (prodComparison F A B)] :
+    (curriedTensor C).obj A ⋙ F ≅ F ⋙ (curriedTensor D).obj (F.obj A) := by
+  refine { @asIso _ _ _ _ _ (?_) with hom := prodComparisonNatTrans F A }
+  apply NatIso.isIso_of_isIso_app
```

Comment:
In order to avoid the `@asIso _ _ _ _ _`, I would define as something like:
```lean
... :=
  letI : IsIso (prodComparisonNatTrans F A) := sorry
  asIso (prodComparisonNatTrans F A)
```

Reply:
Going this way breaks the `@[simps]` so we need to add `_hom` lemmas by hand, and due to some repetitions with below code I figured it’s best to add those instances anyway, but I agree that it’s more satisfying this way.